### ISSUE TITLE
Training / Evaluation Question

### ISSUE THREAD

**Issue by dangthatsright** [cornellius-gp/gpytorch#1201]
Hello, I've been playing around with gpytorch recently and it is great! I just had a few questions around training and evaluation. These are in the context of exact GPs for regression.

I have something like this for training
`
for i in range(training_iter):
    optimizer.zero_grad()
    output = model(train_x)
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   noise: %.3f' % (
        i + 1, training_iter, loss.item(),
        torch.mean(model.likelihood.noise)
    ))
    optimizer.step()
`
and something like this for evaluation
`
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    marginal_likelihood = mll(model(train_x), train_y)
    print(marginal_likelihood.item())
`
These unfortunately, give different values. Is there a reason behind this? Shouldn't they both be capturing the marginal loglikelihood of the training data? 

The related question is how are folks training their GPs? The number of iterations seem arbitrary in the docs, so I was wondering if there was some termination condition people generally use. I sometimes see the loss increasing and oscillating (which is totally fine - if I wanted to treat it as a NN, I would lower learning rate etc), but I figured since GPs have closed forms, there should be a better way to do this rather than hyperparameter tuning the learning rate and early stopping.

**Comment by KeAWang** [comment]
You are likely calling `mll(model(train_x))` after calling `model.eval()` in which case you will get the posterior predictive distribution out of `model(train_x)`. This will be different from the result in `model.train()` model.

In terms of number of iterations, one heuristic is that the MLL is usually pretty calibrated for GPs and a higher likelihood generally leads to better predictive performance. Thus personally I stop the optimization once the marginal likelihood flattens.

**Comment by dangthatsright** [comment]
Thanks for the response! Sorry if I not understanding this properly, but since we are using gradient descent, aren't the posterior and prior distributions near convergence basically the same values? Unless when you say prior you are talking about the initial distribution, but then why would calling `mll(model(train_x))` in train mode use a fixed distribution?

Thanks that makes sense for the stopping criteria, thanks!

**Comment by KeAWang** [comment]
Gradient descent is only changing the hyperparameters of the kernel. It wont cause the prior to be the same as the posterior. Those will always be difference since one is the distribution before seeing data and the other is a distribution conditioned on data.

The marginal log likelihood of a GP is computed from the prior mean and prior covariance matrix among the training points. Hence `mll` expects the prior multivariate normal distribution as its input. Note that you could pass in _any_ multivariate distribution to `mll` and it'll run without errors. This is why you are able to pass in `model(x)`, which is always a multivariate normal distribution, even when `model` is in `.eval()` mode.

**Comment by jacobrgardner** [comment]
(I started writing this before @KeAWang responded again, but thought I'd post it anyways in case it is helpful).

If it helps for concreteness, in `train` mode, `model(train_x)` always returns `MultivariateNormal(0, K)` (here assuming a ZeroMean was used for simplicity). This distribution is not fixed, as it depends on the hyperparameters (e.g., the kernel hypers, the mean constant if there is one), which are changing each iteration of training.

In `eval` mode, `model(x*)` returns `MultivariateNormal(K_{x*X}(K_{XX}+ \sigma^{2} I)^{-1}y, K_{x*x*} - K_{x*X}(K_{XX}+ \sigma^{2} I)^{-1}K_{Xx*})`. When `x* == train_x`.

In most situations excluding the existence of priors, added loss terms for SGPR etc, `mll(some_mvn_distribution, some_output)` is really just a glorified wrapper around `some_mvn_distribution.log_prob(some_output)`.

**Comment by dangthatsright** [comment]
Thanks, these are very helpful! My understanding of the prior and posterior distribution was a flawed. The last question I have is, shouldn't you be stopping training when the mll of the training data with respect to the posterior (and not the prior) distribution plateaus? or is this overfitting?

**Comment by jacobrgardner** [comment]
I would definitely use the mll as output in train mode, as that is the objective function you are training through. I don't think it would make much sense to use your training set for early stopping in eval mode.

I'm going to close this for now, feel free to reopen or open a new issue for further questions.

**Comment by ginward** [comment]
Hello, just another question regarding `mll` and earlystopping. 

If I am using approximate GP and mini-batch training, and if I am supposed to use `mll` in training mode to do earlystopping, how do I add the `mll` of different batches together? Do I just do a plain addition of all the `mll` in each mini-batch?

**Comment by AdrianSosic** [comment]
Hi @jacobrgardner, I have a question concerning the computation of the model's marginal likelihood and stumbled over this post. 

You mentioned that -- without the usage of model priors or SGPR -- the right thing to do in order to get the numerical value of the MLL is to call `mll(model(train_x), train_y)`, which under the hood simply evaluates the log_prob of `train_y` under the multivariate normal that stems from the `model(train_x)` call. 

However, what is the correct approach to evaluate the MLL in the other cases, e.g. when there are priors involved? Based on what I can see in the source code, I guess this somehow works via the magical `params` argument that needs to be passed to the MLL's forward function, which is then used internally to add certain "other terms"? (see snippet below copied from `ExactMarginalLogLikelihood`)
```
# Get the log prob of the marginal distribution
output = self.likelihood(function_dist, *params)
res = output.log_prob(target)
res = self._add_other_terms(res, params)
``` 

Unfortunately, this part is not really documented I wonder with which objects the method would need to be called. Can you give me a hint?
